wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeInfrastructure.java

```diff
@@ -143,4 +155,93 @@ protected abstract void internalEstimate(InternalEnvironment env)
+
+  /**
+   * Resolves {@link InternalEnvironment} instance based on specified {@link Environment}.
+   *
+   * <p>Resolved internal environment will have:
+   *
+   * <ul>
+   *   <li>Downloaded recipe;
+   *   <li>Fetched information about configured {@link Installer installers};
+   *   <li>Full servers list that includes: configured servers by users and provided by installers.
+   * </ul>
+   *
+   * @param environment environment to resolve
+   * @return resolved internal environment
+   * @throws InfrastructureException if any exception occurs on environment resolving
+   * @see InternalEnvironment
+   */
+  private InternalEnvironment resolveInternalEnvironment(Environment environment)
+      throws InfrastructureException {
+    InternalRecipe internalRecipe = recipeRetriever.getRecipe(environment.getRecipe());
+
+    Map<String, InternalMachineConfig> internalMachines = new HashMap<>();
+    for (Map.Entry<String, ? extends MachineConfig> machineEntry :
+        environment.getMachines().entrySet()) {
+      MachineConfig machineConfig = machineEntry.getValue();
+      List<Installer> installers = getInstallers(machineConfig.getInstallers());
+
+      Map<String, ServerConfig> servers = new HashMap<>(machineConfig.getServers());
+
+      fillServers(servers, installers);
+
+      servers = normalizeServers(servers);
+
+      internalMachines.put(
+          machineEntry.getKey(),
+          new InternalMachineConfig(
+              installers, servers, machineConfig.getEnv(), machineConfig.getAttributes()));
+    }
+    return new InternalEnvironment(internalRecipe, internalMachines);
+  }
+
+  private List<Installer> getInstallers(List<String> installersKeys)
+      throws InfrastructureException {
+    try {
+      return installerRegistry.getOrderedInstallers(installersKeys);
+    } catch (InstallerException e) {
+      throw new InfrastructureException(e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Fill specified map with servers that are provided by installers.
+   *
+   * @param servers map to fill
+   * @param installers installers to retrieve servers
+   * @throws InfrastructureException if any installer has server that conflicts with already
+   *     configured one
+   */
+  private void fillServers(Map<String, ServerConfig> servers, List<Installer> installers)
+      throws InfrastructureException {
+    for (Installer installer : installers) {
+      for (Map.Entry<String, ? extends ServerConfig> serverEntry :
+          installer.getServers().entrySet()) {
+        if (servers.putIfAbsent(serverEntry.getKey(), serverEntry.getValue()) != null
+            && !servers.get(serverEntry.getKey()).equals(serverEntry.getValue())) {
+          throw new InfrastructureException(
+              String.format(
+                  "Installer '%s' contains server '%s' conflicting with machine configuration",
+                  installer.getId(), serverEntry.getKey()));
+        }
+      }
+    }
+  }
+
+  //TODO Take a look | Add tests
```

Comment:
Are you going to add tests in this PR?

Reply:
It is already fixed by https://github.com/eclipse/che/pull/6983/commits/b39856235c7b33d839e9f6258d433e6721eb3c9f